README.md
# Greedy Maximum Entropy Sampler

The objective is to extract a sample $S$ of items from an initial set $D$ with an optimized diversity of the attached entities: $S \subset D$, $|S| = l$ and $|D| = L$.

In the initial set $D$, each document (or item) $d$ is attached with a set of $N$-ary relations: $\{r_1, r_2, ..., r_{n_d}\}^{d}$, where $n_d$ in the number of relations attached to $d$. Each relation $r$ is connecting a set of $N$ entities, $r=(e_1, e_2, e_3, e_N)$, each belonging one particular *category* represented by a larger set and denoted by the variables $V_1, V_2, ..., V_{N}$. Sets are considered distincts and each entity belongs to one set: $e \in V_1 \cup V_2 ... \cup V_N$ and $\forall i, j, V_i \cap V_j = \emptyset$.


Also, one entity $e_i$ can appear many times in $d$: $0 \le |\{ r_k: e_i \in r_k\}^{d}| \le n_d$, with $|\{r_k: e_i \in r_k \}^{d}| \text{ the number of }N \text{-ary relations involving }e_i\text{  in }d\text{ .}$

Then, given a subset $S$, the probability that a $N$-ary relation involve the entity $e_i$ is:

$P(e_i) = \frac{\displaystyle\sum_{d \in S} |\{ r_k: e_i \in r_k \}^{d}|}{\displaystyle\sum_{d \in S} n_d}$


It follows that the diversity of organisms or chemicals can be measured with the Shannon's entropy over the probability distributions of elements $V_{1:N}$. Expressed with entropy, the diversity reflects the uncertainty about the entities attached to a relation in an document $d$. The Shannon's entropy in the sample $S$ for the set $V_j$ is $H_S(V_j) = - \displaystyle\sum_{e_i \in V_i} P(e_i) \log P(e_i)$. Adding a new document $\displaystyle d$ to $S$ will update the probability distributions of $V_{1:N}$, and the new observed entropy of $V_j$ for instance will be $H_{\displaystyle S_{+d}}(V_j)$ where $\displaystyle S_{+d} = S \cup \{d\}$. Therefore, to optimize the diversity over $V_{1:N}$, the document $d*$ added to $S$ minimizes the distance to the utopian point $(log |V_1|, log |V_2|, ..., log |V_N|)$ (maximal observable entropies):

$d* = \underset{d}{argmin} || (H_{\displaystyle S_{+d}}(V_1), H_{\displaystyle S_{+d}}(V_2), ..., H_{\displaystyle S_{+d}}(V_N)) - (log |V_1|, log |V_2|, ..., log |V_N|) ||.$

The proposed sampling approach is a simple greedy algorithm that, at each step, selects and adds the new document $d*$ from $D$, maximising the diversity over the different set of entities  $V_{1:N}$. We refer to it as diversity-sampling. As in ecology, diversity is intrinsically related to the number of distinct entities per each set, but, for the purpose of using the samples as training or evaluation sets, the balance of the distribution is prioritized over a higher number of rare entities. This latter behaviours is a natural consequence of using the Shannon's entropy. From this perspective, the method can also been seen as a ranking procedure, and a sample is determined by selecting the first top $n$ ranked items. The selection of an appropriate sample size $l$ is also a critical, but often overlooked factor. By monitoring $H_S(V_{j})$ $\forall j \in 1:N$ during the iterative construction of $S$ (until $l=L$), it is possible to determine the step $l$ at which diversity starts to deteriorate and sampling should be stopped, i.e. when the new added documents provide relationships for already frequently reported entities in $S$.



## Install

Install the gme-sampler
```bash
pip install git+ssh://git@github.com/idiap/gme-sampler
```

## Usage

The provided example considers the motivating scenario of sampling literature references for the LOTUS database. 

The original snapshot of the LOTUS database (v.10 - Jan 6, 2023) used in this work is available at 
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7534071.svg)](https://doi.org/10.5281/zenodo.7534071).

The LOTUS dataset used is licensed under CC BY 4.0. See the related [article](https://elifesciences.org/articles/70780) and the [website](https://lotus.naturalproducts.net/) for more details.

Only a subset of the original database is used in this example.
In this example, each document $d$ is identified by the *reference_doi* columns, e.g: [10.1002/(SICI)1099-1573(199902)13:13.0.CO;2-F](https://onlinelibrary.wiley.com/doi/10.1002/(SICI)1099-1573(199902)13:1%3C75::AID-PTR387%3E3.0.CO;2-F). Each document then reports a list of relations between organisms and chemicals, respectively identified by a Wikidata Organisms entity and a Wikidata structure entity. In this example, we sample to top-5 documents that maximise the diversity among organisms and chemicals from this subset. The dataset looks like this:



| structure_wikidata                        | structure_cid | structure_nameTraditional | organism_wikidata                       | organism_name     | organism_taxonomy_02kingdom | reference_wikidata                       | reference_doi                                 | reference_pubmed_id |
|-------------------------------------------|---------------|---------------------------|-----------------------------------------|-------------------|-----------------------------|------------------------------------------|-----------------------------------------------|---------------------|
| http://www.wikidata.org/entity/Q43656     | 5997          | Cholesterol               | http://www.wikidata.org/entity/Q1146782 | Eryngium foetidum | Archaeplastida              | http://www.wikidata.org/entity/Q34502919 | 10.1002/(SICI)1099-1573(199902)13:13.0.CO;2-F | 10189959            |
| http://www.wikidata.org/entity/Q121802    | 222284        | Beta-Sitosterol           | http://www.wikidata.org/entity/Q1146782 | Eryngium foetidum | Archaeplastida              | http://www.wikidata.org/entity/Q34502919 | 10.1002/(SICI)1099-1573(199902)13:13.0.CO;2-F | 10189959            |
| http://www.wikidata.org/entity/Q104253515 | 5283638       | Clerosterol               | http://www.wikidata.org/entity/Q1146782 | Eryngium foetidum | Archaeplastida              | http://www.wikidata.org/entity/Q34502919 | 10.1002/(SICI)1099-1573(199902)13:13.0.CO;2-F | 10189959            |


- The *item_column* indicates the column containing the documents identifiers while columns containing the variables to maximise the diversity are specified with the *on_columns* argument.

- If *binarised* is set to True, thenwe only consider the distinct set of entities per documents.

- *dutopia* is the main metric use to select the next document are represent the distance to the uptopian point, see [here](#greedy-maximum-entropy-sampler). However, a second metric is provided, *sum*, which simply rank the documents by the sum of the associated entropies.

```python
import pandas as pd
from gme.gme import GreedyMaximumEntropySampler

# Load data
data = pd.read_csv("data/test_data.csv", sep="\t", dtype=object)

# Init sampler
sampler = GreedyMaximumEntropySampler(selector="dutopia", binarised=False)

# Sample
output = sampler.sample(
        data=data,
        N=5,
        item_column="reference_doi",
        on_columns=["structure_wikidata", "organism_wikidata"],
    )
```

The expected output is:

|         reference_doi         | structure_wikidata | organism_wikidata |
|:-----------------------------:|:------------------:|:-----------------:|
|  10.1016/0039-128X(82)90018-6 |       2.77259      |      0.00000      |
|      10.1055/S-2001-11496     |       2.89793      |      1.02910      |
| 10.1016/J.TALANTA.2005.04.043 |       3.32340      |      1.33408      |
|  10.1016/0039-128X(80)90068-9 |       3.57149      |      1.56290      |
| 10.1016/S0305-1978(00)00054-5 |       3.70730      |      1.75229      |

At each step the document which maximised the diversity is sampled and the corresponding increasing entropy values for each variables (here *structure_wikidata* and *organism_wikidata*) are indicated.

On larger datasets, monitoring the entropy values can help estimate a sufficient sample size. See a complete example in our related [article](http://arxiv.org/abs/2311.06364).

## Additional information

The approache scales quadratically with the size of the dataset, which can be impractical for very large datasets. In this context, we implemented an ```--approx``` option. With this option, instead of computing all the emtropy values, the best candidate is approximated by only taking the best one over a random sample of $m$ items. If 0 (default), no approximation is performed.

Considering an intial sample of $N$ items and a used (*approx*) sample size of $k$, we can estimate the probability that the document $d*$ selected (top-1 in the sample of size $k$) is at least in the top-$n$ in the full dataset, with: $1 - [{(N - n) \choose k} / {N \choose k}]$.

For instance, for a dataset of size $N=130,000$, with a sample for *approx* of size $25,000$ (reduced by a factor $5$), we can estimate that there is:
- 88.19 % that it is in the top-10 in the full dataset.
- 98.60 % that it is in the top-20 in the full dataset.
- 99.99 % that it is in the top-50 in the full dataset.
- $\approx 100$ % that it is in the top-100 in the full dataset.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - Date

First official release!

### Added

- Feature 1
- Feature 2


tests/test_gme.py
# SPDX-FileCopyrightText: 2023 Idiap Research Institute <contact@idiap.ch>
#
# SPDX-FileContributor: Delmas Maxime maxime.delmas@idiap.ch
#
# SPDX-License-Identifier: GPL-3.0-or-later

import pandas as pd
from pandas.testing import assert_frame_equal

from gme.gme import GreedyMaximumEntropySampler


def test_sampling_with_dummy_data_correct_outputs():
    # Load dummies data
    data = pd.read_csv("data/test_data.csv", sep="\t", dtype=object).fillna("")
    expected_output = pd.DataFrame(
        {
            "reference_doi": [
                "10.1016/0039-128X(82)90018-6",
                "10.1055/S-2001-11496",
                "10.1016/J.TALANTA.2005.04.043",
                "10.1016/0039-128X(80)90068-9",
                "10.1016/S0305-1978(00)00054-5",
                "10.1016/0166-6851(85)90115-X",
                "10.1007/BF02535649",
                "10.1007/BF02879004",
                "10.1021/JF60213A026",
                "10.1002/(SICI)1099-1573(199902)13:1<78::AID-PTR384>3.0.CO;2-F",
            ],
            "structure_wikidata": [
                2.77259,
                2.89793,
                3.3234,
                3.57149,
                3.7073,
                3.70688,
                3.8136,
                3.89296,
                3.80794,
                3.78291,
            ],
            "organism_wikidata": [
                0.0,
                1.0291,
                1.33408,
                1.5629,
                1.75229,
                1.96591,
                2.09647,
                2.20116,
                2.37203,
                2.4612,
            ],
        }
    )

    # eval outputs
    sampler = GreedyMaximumEntropySampler(selector="dutopia", binarised=False)
    output = sampler.sample(
        data=data,
        N=10,
        item_column="reference_doi",
        on_columns=["structure_wikidata", "organism_wikidata"],
    )

    assert_frame_equal(output, expected_output)


tests/test_helpers.py
# SPDX-FileCopyrightText: 2023 Idiap Research Institute <contact@idiap.ch>
#
# SPDX-FileContributor: Delmas Maxime maxime.delmas@idiap.ch
#
# SPDX-License-Identifier: GPL-3.0-or-later

import numpy as np
from numpy.testing import assert_array_equal

from gme.helpers import compute_all_entropy
from gme.helpers import compute_entropy
from gme.helpers import indexes_to_vector


def test_indexes_to_vector_returns_correct_output():
    out = indexes_to_vector([1, 1, 2, 2, 4], 10)
    assert_array_equal(out, np.array([0, 2, 2, 0, 1, 0, 0, 0, 0, 0]))


def test_compute_entropy_returns_correct_outputs():
    out = compute_entropy([[1, 2, 3, 4], [1, 2]], [0, 0, 1, 2, 0])
    assert out == [1.27703, 1.05492]


def test_compute_all_entropies_correct_outputs():
    out = compute_all_entropy(
        [[[1, 2, 3, 4], [1, 2]], [[4, 3, 2, 1], [2, 1]]],
        [[0, 0, 1, 2, 0], [0, 2, 1, 0, 0]],
    )
    assert out == [[1.27703, 1.05492], [1.27703, 0.67301]]


gme/helpers.py
# SPDX-FileCopyrightText: 2023 Idiap Research Institute <contact@idiap.ch>
#
# SPDX-FileContributor: Delmas Maxime maxime.delmas@idiap.ch
#
# SPDX-License-Identifier: GPL-3.0-or-later

"""helpers functions"""
import itertools
import logging
import random
import sys
from multiprocessing import Pool
from typing import List

import numpy as np
import pandas as pd
from scipy.stats import entropy


def set_seed(seed):
    """
    Set the random seed.
    """
    random.seed(seed)
    np.random.seed(seed)


def read_data(path: str, sep: str, logger: logging.Logger) -> pd.DataFrame:
    """Helper to read a dataframe

    Args:
        path (str): path to the file
        sep (str): separator
        logger (logging.Logger): logger instance

    Returns:
        pd.DataFrame: the dataframe
    """

    logger.info("Read file at %s", path)

    try:
        data = pd.read_csv(path, sep=sep, dtype=object).fillna("")

    except pd.errors.ParserError as except_parsing_error:
        logger.error(
            "file at %s has incorrect format. %s", path, str(except_parsing_error)
        )
        sys.exit(1)

    except FileNotFoundError:
        logger.error("File not found at %s", path)
        sys.exit(1)

    return data


def indexes_to_vector(indexes: List[int], n: int) -> np.ndarray:
    """Convert a vector of indexes to a vector of counts.
    The vector is initialized as a zero-vector of size n.
    The number of times the index i appears in the given list
    of indexes is then used to calculate the value at position i.

    Args:
        indexes (list): the list of indexes
        n (int): the size to give to the vector

    Returns:
        _type_: the vector ()
    """
    if max(indexes) >= n:
        raise ValueError("Maximum index is bigger than size of vector")

    v = np.zeros(n, dtype=np.int32)
    for i in indexes:
        v[i] += 1
    return v


def compute_entropy(indexes_per_doi: list, current_vector: np.array) -> np.array:
    """For each list of indexes provided in indexes_per_doi, compute the entropy that
    would correspond to the addition of these counts to the current entropy vector.
    Recall that each list of indexes correspond the list of involved entities.

    Args:
        indexes_per_doi (list): a list of list of indexes, e-g: [ [1, 4, 60],  [24, 65, 98, 34], ..., [15, 23, 98]]
        current_vector (np.array): the current vector of counts. Each entry correspond to the current number
        of relation involving the ith entity identified by its index.

    Returns:
        np.array: _description_
    """

    n = len(current_vector)
    h_vector = []

    for indexes in indexes_per_doi:
        probs = indexes_to_vector(indexes, n)
        probs += current_vector
        probs = probs / probs.sum()
        e = entropy(probs)
        h_vector.append(np.round(e, decimals=5))

    return h_vector


def compute_all_entropy(l_indexes_per_doi: list, l_occurence_v: list):
    """Giving a list containing a list of vectors for each set of entities,
    along with a list containing the current occurence vector of each set of entities,
    call the entropy function for all of them sequentially.

    Args:
        l_indexes_per_doi (list): [list of list of indexes for each entity type]
        Eg. [[[1,2,3,4], [1,2]], [[4,3,2,1], [2,1]]] where [1,2,3,4] is the set of
        indexes of entities associated with the item 1 in the first variable, and respectively,
        [2,1] is the set of indexes of entities associated with the second variable
        for the second item.
        l_occurence_v (list): [current occurence vector for for each entity type]
        E.g [[0, 0, 1, 2, 0], [0, 2, 1, 0, 0]] The entity at index 2 of the first variable already
        happened 2 times in the current set, while entity at index 0 for instance in the
        second variable do not appears yet.

    Returns:
        _type_: a list of the list of computed entropies
    """
    out = []

    if not len(l_indexes_per_doi) == len(l_occurence_v):
        raise ValueError(
            "Different sizes of the current running occurence vector and the one containing the indexes per items"
        )

    n = len(l_indexes_per_doi)

    for i in range(n):
        next_h = compute_entropy(l_indexes_per_doi[i], l_occurence_v[i])
        out.append(next_h)

    return out


def distribute_entropy_on_workers(
    list_of_indexes_per_item: list, list_of_occurences: list, n_worker: int
) -> list:
    """Parrelise the computation of entropies by ditributing the list of indexes to different workers

    Args:
        list_of_indexes_per_item (list): a list of list of indexes of entities for each variable per DOI
        list_of_occurences (list): the current occurence vector for each variable
        n_worker (int): _description_

    Returns:
        list: list of numpy.arrays containing entropy values.
    """

    if not all(
        [
            len(l_item) == len(list_of_indexes_per_item[0])
            for l_item in list_of_indexes_per_item
        ]
    ):
        raise ValueError("all the list don't have the same size !")
    n = len(list_of_indexes_per_item[0])

    if n <= n_worker or n_worker == 1:
        res = compute_all_entropy(list_of_indexes_per_item, list_of_occurences)
        out = [np.array(x) for x in res]

        return out

    workers_ranges = [(a[0], a[-1]) for a in np.array_split(list(range(n)), n_worker)]

    splited_list_of_indexes_per_item = [
        [l_indexes[r[0] : r[1] + 1] for r in workers_ranges]
        for l_indexes in list_of_indexes_per_item
    ]

    # To send to the worker, we need to split the data, so that each one can working individually in parallel.
    # params in in the form of: [ worker_1, worker_2, ..., worker_n ]. Earch worker in then a list of the parameters
    # that have to be send to compute_all_entropy, so worker_i = [l_indexes_per_doi, l_occurence_v]
    # Then, l_indexes_per_doi is a list, containing the list of indexes for orgamisms and chemical, in the form of:
    #  l_indexes_per_doi = [ [l_o1, l_o2, l_o3, ..., l_on] , [l_c1, l_c2, l_c3, ..., l_cn] ].
    # here n give the nb of example gave to this worked.
    # l_occurence_v in the a list containing the both vector v_org and v_chemical
    # Recall that l_o1 for instance is by itself a list containg all the indexes of the organisms
    # assocaited to this DOI.
    params = list(
        zip(
            (list(zip(*splited_list_of_indexes_per_item))),
            [list_of_occurences for i in range(n_worker)],
        )
    )

    with Pool(processes=n_worker) as pool:
        res = pool.starmap(compute_all_entropy, params)

    out = [list(itertools.chain.from_iterable(r)) for r in zip(*res)]
    out = [np.array(x) for x in out]

    return out


gme/gme.py
# SPDX-FileCopyrightText: 2023 Idiap Research Institute <contact@idiap.ch>
#
# SPDX-FileContributor: Delmas Maxime maxime.delmas@idiap.ch
#
# SPDX-License-Identifier: GPL-3.0-or-later

"""GME sampler"""
import gc
import logging
import os
import random
import sys

import numpy as np
import pandas as pd

from gme.helpers import distribute_entropy_on_workers
from gme.helpers import indexes_to_vector


class GreedyMaximumEntropySampler:
    """The GME class"""

    def __init__(self, selector: str, binarised: bool, **kwargs):
        self.selector = selector
        self.binarised = binarised
        self.logger = kwargs.pop("logger", self.get_std_logger())
        self.n_workers = kwargs.pop("n_workers", 1)

    def get_std_logger(self):
        """Create a default logger

        Returns:
            logging.logger: a default logger
        """

        # set loggers
        log_path = os.path.join(
            "./sampling_GME_" + ("_freq" if not self.binarised else "") + ".log"
        )
        open(log_path, "w", encoding="utf-8").close()
        handlers = [
            logging.FileHandler(filename=log_path),
            logging.StreamHandler(stream=sys.stdout),
        ]
        logging.basicConfig(
            handlers=handlers,
            format="%(asctime)s %(levelname)-8s %(message)s",
            level=logging.DEBUG,
        )
        logger = logging.getLogger("lotus-dataset-creator")

        return logger

    def sample(
        self,
        data: pd.DataFrame,
        N: int,
        item_column: str,
        on_columns: list,
        approx: int = 0,
    ):
        """Create a sample.
        The sampled elements are individual instance from 'item_column'.
        Each item will be select to maximise the entropy its related entites in the 'on_columns' attributes.

        Args:
            N (int): number of item from 'item_column' to sample
            item_column (str): individual column
            on_columns (list): list of columns name for the attributes to maximise entropy on.
            approx (int): Instead of computing all the emtropy values, approximate the best candidate by
              only taking the best one over a random sample of n items. If 0, no approximation.
        """

        def preprocess(data, on_columns: list):
            """Preprocess the table to attribute a unique index to each entities in the 'on_columns' attributes.

            Args:
                data (pd.DataFrame): the target dataFrame, containing the 'on_columns' attributes.
                on_columns (list): list of columns name for the attributes to maximise entropy on.

            Returns:
                processed_data (pd.DataFrame): the processed dataframe with new columns added "X_indexes", indicating
                the index of each entity in its corresponding set.
            """

            def join_indexes(data: pd.DataFrame, column: str) -> pd.DataFrame:
                """Determine the indexes for each distinct elements in a column.

                Args:
                    data (pd.DataFrame): the dataframe
                    column (str): the targeted column, e.g: organism_wikidata or structure_wikidata
                    in the working example.

                Returns:
                    pd.DataFrame: the dataframe with a new column adding a unique index for each distinct entity
                    in the column.
                """
                self.logger.info("Preprocessing column %s", column)
                list_of_entities = list(set(data[column]))
                entities_to_join = pd.DataFrame(
                    {
                        column: list_of_entities,
                        column + "_indexes": range(len(list_of_entities)),
                    }
                )
                return data.merge(entities_to_join, how="left", on=column), len(
                    list_of_entities
                )

            n_items_per_columns = []

            for column in on_columns:
                data, n = join_indexes(data, column)
                n_items_per_columns.append(n)

            return data, n_items_per_columns

        def selector(s_type: str, next_h_values: list, n_items_per_columns) -> int:
            """The selector for the next article to add to the dataset.

            Args:
                s_type (str): the seldctor type
                next_h_values (list): list of np.array of new entropies on columns attributes
                n_items_per_columns (list): list of the cardinality (number of distinct items) per
                columns attributes

            Raises:
                ValueError: an error

            Returns:
                int: the index of the new DOI that maximise the chosen selection criteria.
            """
            agg_type = ["sum", "dutopia"]

            if s_type not in agg_type:
                self.logger.error(
                    "Invalid aggregator type. Expected one of: %s", ",".join(agg_type)
                )
                raise ValueError("Invalid aggregator type")

            if s_type == "sum":
                s = sum(next_h_values)
                return s.argmax()

            if s_type == "dutopia":
                # Dist in the norm of the diff
                coord = np.array(next_h_values)

                diff = coord.T - np.array(
                    [np.log(nb_items) for nb_items in n_items_per_columns]
                )
                d_utopia = np.linalg.norm(diff, axis=1)

                return d_utopia.argmin()

        # Copy of working data
        working_data = data.copy()
        # preprocess the data
        working_data, n_items_per_columns = preprocess(working_data, on_columns)
        n_cols = len(on_columns)

        # Only consider the set of elements if binarised, not the freq
        list_of_indexes_per_item = []
        if self.binarised:
            for column in on_columns:
                list_of_indexes_per_item.append(
                    working_data.groupby(item_column)[column + "_indexes"]
                    .apply(set)
                    .apply(list)
                )
        else:
            for column in on_columns:
                list_of_indexes_per_item.append(
                    working_data.groupby(item_column)[column + "_indexes"].apply(list)
                )

        if not all(
            [
                list(c.index) == list(list_of_indexes_per_item[0].index)
                for c in list_of_indexes_per_item
            ]
        ):
            raise ValueError("Indexes per %s are not in the same order" % (item_column))

        # Extract the list of item. It has been assert that it is in the same order for all the 'on_columns' attributes
        list_of_items = list_of_indexes_per_item[0].index.to_list()

        if N > len(list_of_items):
            self.logger.warning(
                "N=%d greater than the total number of available DOI (%d). Decreasing N to the max size.",
                N,
                len(list_of_items),
            )
            N = len(list_of_items)

        # Transform this in a list
        list_of_indexes_per_item = [
            indexes_per_item.tolist() for indexes_per_item in list_of_indexes_per_item
        ]
        list_of_occurences = [
            np.zeros(n_items, dtype=np.int32) for n_items in n_items_per_columns
        ]

        # init
        h_values = [0 for i in range(len(on_columns))]
        sampled_items = []
        l_iter_h = [[] for i in range(len(on_columns))]

        # loop until N examples have been selected
        while len(sampled_items) < N:
            # Check that the list always have the same
            t = [len(l_items) for l_items in list_of_indexes_per_item]
            t.append(len(list_of_items))
            if not len(set(t)) == 1:
                raise ValueError(
                    "Different number of items between individual items between columns."
                )

            if approx < len(list_of_items) and approx > 0:
                # If an approx sample size was provided, only computed Entropy values on a random sample
                sample_indexes = random.sample(range(len(list_of_items)), approx)
                working_list_of_indexes_per_item = [
                    [list_of_indexes_per_item[i][j] for j in sample_indexes]
                    for i in range(n_cols)
                ]

                # Step 1: Compute all possible entropy
                next_h_values = distribute_entropy_on_workers(
                    working_list_of_indexes_per_item, list_of_occurences, self.n_workers
                )
                gc.collect()

                # Step 2: get the doi maximazing H according to the defined selector
                working_selected = selector(
                    self.selector, next_h_values, n_items_per_columns
                )
                selected = sample_indexes[working_selected]
                item = list_of_items.pop(selected)

                # Extarct corresponding H value and compare to previous. Log if needed (On sample !)
                new_h_values = [
                    l_h_values[working_selected] for l_h_values in next_h_values
                ]

            else:
                # Step 1: Compute all possible entropy
                next_h_values = distribute_entropy_on_workers(
                    list_of_indexes_per_item, list_of_occurences, self.n_workers
                )
                gc.collect()

                # Step 2: get the doi maximazing H according to the defined selector
                selected = selector(self.selector, next_h_values, n_items_per_columns)
                item = list_of_items.pop(selected)

                # Extarct corresponding H value and compare to previous. Log if needed
                new_h_values = [l_h_values[selected] for l_h_values in next_h_values]

            line = "| {:<50} | {:>10}".format(
                item,
                "".join(
                    [
                        "H(" + on_columns[i] + ") = " + f"{new_h_values[i]:.2f}" + " | "
                        for i in range(n_cols)
                    ]
                ),
            )
            self.logger.info("%s", line)

            for i in range(n_cols):
                if new_h_values[i] < h_values[i]:
                    self.logger.warning(
                        "Adding %s decreased H(%s): %.2f -> %.2f",
                        item,
                        on_columns[i],
                        h_values[i],
                        new_h_values[i],
                    )

            # Add the sampled item to the list
            sampled_items.append(item)

            # For all columns attribute
            for i in range(n_cols):
                # Add the new H value
                l_iter_h[i].append(new_h_values[i])

                # Extract the vector, add it to the current occurence vector and remove
                #  the corresponding line in the crosstab
                list_of_occurences[i] += indexes_to_vector(
                    list_of_indexes_per_item[i].pop(selected), n_items_per_columns[i]
                )

                # update new H values
                h_values[i] = new_h_values[i]

        df_H = pd.DataFrame(
            dict(zip(([item_column] + on_columns), ([sampled_items] + l_iter_h)))
        )

        return df_H
